Rewrite primitives exercise (#55)
component-driven/component-driven-development: src/exercises/3-Primitives/Readme.md

In this exercise we’ll create reusable primitive components. We’ll create a text and a heading component. We'll learn how to reduce boilerplate by using [styled-system](https://styled-system.com)

> About 80% of every user interface is a text. Unsurprisingly, most of inconsistencies are coming from text styles. In order to reduce the amount of different text styles we should restrict our styling and only allow using design tokens.

Reviewer: ```suggestion
The majority of every user interface is a text. Unsurprisingly, most of inconsistencies are coming from text styles. To reduce the amount of different text styles we should restrict our styling and only allow using design tokens.
```

Otherwise we'll need to provide the source of this data.

## 3.1. Creating a generic text component

In the previous exercise we learned how to use design tokens instead of arbitrary CSS values in our components. Now we need a way to write styles and access design tokens inside our application code. I.e. we want to design a component that implements an API to our design system and doesn't require writing CSS.

We could approach this task naively:

```jsx static
const Text = styled.p`

>   fontfamily: ${props => props.theme.fonts[props.fontFamily]};

Reviewer: ```suggestion
  font-family: ${props => props.theme.fonts[props.fontFamily]};
```

>   fontsize: ${props => props.theme.fontSizes[props.fontSize]};

Reviewer: ```suggestion
  font-size: ${props => props.theme.fontSizes[props.fontSize]};
```

>   fontweight: ${props => props.theme.fontWeights[props.fontWeight]};

Reviewer: ```suggestion
  font-weight: ${props => props.theme.fontWeights[props.fontWeight]};
```

When working on the app or a website it's oftentimes desirable to handle responsive styles as well. There are different ways of handling responsive styles. Styled-system approach is simple yet very powerful: every prop accepts a value or an array of values.

### The task

- Make the font size of `Heading` component responsive. It should become smaller, on smaller screens.

**Hint:** Check out the documentation of [responsive styles](https://styled-system.com/responsive-styles)

## 3.5 Making primitives lean

It is not possible to account for all use cases for your design system. As with any software, requirements are going to change over time and it is crucial for primitives to be lean enough to adapt to those requirement. In other words, good primitives should be flexible enough to allow one-off "snowflakes" usages of unpredicted requirements.

### The task

- Refactor `Text` component so it could accept style overrides via `css` prop. All values should accept design tokens and not just CSS values.

**Hint:** Check out how to use [CSS prop](https://www.styled-components.com/docs/api#css-prop) in styled-components and [documentation of CSS function](https://styled-system.com/css)

> ## 3.6 Adding responsive styles via primitives

Reviewer: Copypaste from 3.4?

Author: Good catch! Forgot to remove

<Spacing theme={theme} />;
```

## Solutions

<details>
 <summary>Text</summary>

```js {"file": "final/Text.js", "static": true}
```

</details>

<details>
 <summary>Heading</summary>

```js {"file": "final/Heading.js", "static": true}
```

</details>

## Takeaways

1. All text in the UI should be rendered using `Text` component
1. Do not rely on global CSS since it is error prone and may differ depending on where the component is rendered.
1. Use design tokens from the design system to reduce the amount of choice.
1. When creating primitives, first think of the API (i.e. imagine how the component is going to be used, then code it).

> 1. styled-system allows creating primitives with constraint-based APIs easily.

Reviewer: ```suggestion
1. styled-system simplifies creating primitives with constraint-based APIs.
```

> 1. Compose higher order components out of primitives.

Reviewer: ```suggestion
1. Compose complex components out of primitives.
```

HoC are a particular pattern, which isn't what we're doing here.
